Throw an exception when calling `MembershipManager::getGroupIds()` with an entity which is not group content

pfrenssen opened the pull request:
While working on a test for #296 I noticed that `og_invalidate_group_content_cache_tags()` calls `MembershipManager::getGroupIds()` even when the entity is a group and not group content. This is useless, `::getGroupIds()` only works on group content and is clearly documented as such.

This results in a useless method being called and its useless results cached every time one of the entity hooks is firing on a group entity. Let's fix the bug and also throw an exception when the wrong entity is being passed, so that developers are alerted if they are using the method incorrectly.

pfrenssen committed:
Make sure that MembershipManager::getGroupIds() is only called for group content entities.

pfrenssen committed:
Don't check if OgMembership entities are associated with groups before deleting them.

This check only makes sense for group content entities.

pfrenssen committed:
Don't request groups for entities that are not group content.

This does a needless query, and now also throws an exception.

pfrenssen commented:
The schema errors are possibly due to https://www.drupal.org/node/2870971.

pfrenssen commented on og.module:
```diff
@@ -289,10 +289,9 @@ function og_entity_type_alter(array &$entity_types) {
+  // group listings to be cached effectively. The cache tag format is
   // 'og-group-content:{group entity type}:{group entity id}'.
-  if (Og::isGroup($entity->getEntityTypeId(), $entity->bundle()) || Og::isGroupContent($entity->getEntityTypeId(), $entity->bundle())) {
+  if (Og::isGroupContent($entity->getEntityTypeId(), $entity->bundle())) {
```
This check to see whether the entity is a group is useless, in `MembershipManager::getGroupIds()` the queries are performed on group audience fields. If there are no group audience fields it won't return any results.

pfrenssen commented on src/OgAccess.php:
```diff
@@ -262,36 +262,37 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
-    // The entity might be a user or a non-user entity.
-    $groups = $entity->getEntityTypeId() == 'user' ? $this->membershipManager->getUserGroups($entity) : $this->membershipManager->getGroups($entity);
-
-    if ($is_group_content && $groups) {
```
We already know at this point whether the entity is group content, so a small reshuffle of these two lines into the if statement is sufficient to not do an unneeded call to `MembershipManager::getGroups()`.

pfrenssen commented on src/OgDeleteOrphansBase.php:
```diff
@@ -135,10 +146,17 @@ protected function deleteOrphan($entity_type, $entity_id) {
-    $group_count = $this->membershipManager->getGroupCount($entity);
-    if ($group_count == 0) {
+    if ($this->groupAudienceHelper->hasGroupAudienceField($entity->getEntityTypeId(), $entity->bundle())) {
+      $group_count = $this->membershipManager->getGroupCount($entity);
```
Only do a group count if the entity is actually group content.

MPParsley replied:
```suggestion
      // Only do a group count if the entity is actually group content.
      $group_count = $this->membershipManager->getGroupCount($entity);
```

pfrenssen commented on config/schema/og.schema.yml:
```diff
@@ -153,3 +153,7 @@ condition.plugin.og_group_type:
+
+entity_reference_selection.og:default:
+  type: entity_reference_selection.default
+  label: 'OG entity selection handler settings'
```
Schema change required due to https://www.drupal.org/node/2870971 - I can split this off to a separate ticket if required.

pfrenssen replied:
Let's try leaving this out, this hasn't proven to be problematic in the last 10 months so it was maybe a temporary hiccup that is no longer needed.

pfrenssen commented on og.module:
```diff
@@ -317,8 +317,7 @@ function og_entity_type_alter(array &$entity_types) {
   // tags for each of the groups this group content belongs to. This allows
-  // group listings to be cached effectively. Also invalidate the tags if the
-  // group itself changes. The cache tag format is
+  // group listings to be cached effectively. The cache tag format is
```
This change should be undone, the code is now effectively checking if the group itself changes, this was added in #337.

pfrenssen commented:
Haha Github doesn't allow me to reject my own changes :grin:

pfrenssen commented:
The schema change can just be removed, it has already been fixed in #323 by @claudiu-cristea. It is now duplicating it. I will clean this up and then rebase to clean up the commit history.

pfrenssen commented:
OK this is looking good now. Some of the changes that were included here have been solved in other PRs in the meantime, so It is easier to read. It should be pretty easy to review now.

tatarbj commented:
For me it seems a bit over of the original scope, maybe renaming the PR would make sense. Apart of it as we've also discussed there are no other cases when deleteOrphan() will be executed so it seems safe to delete the entity when it's not a group content (else case).
It looks good to me, thanks @pfrenssen for the clear comments, helped me a lot of understand the cases, also the inline comments are clear and proper!

pfrenssen commented:
Do you mean this comment?

```
+    // If the entity is not group content (e.g. an OgMembership entity), just go
+    // ahead and delete it.
```

That is already what was happening before this PR, I just added some documentation here to explain what is going on.

I don't see where this is going out of scope? The PR adds the new exception that is being thrown if the `getGroupIds()` method is called with a wrong argument. There were two instances where this was called with the wrong argument in our own code, and these two are now fixed by first checking that the entity is group content before calling the method. And a test is added. There is nothing else in the PR?

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into reject-non-group-content

pfrenssen committed:
Merge remote-tracking branch 'gizra/8.x-1.x' into reject-non-group-content

MPParsley commented on src/MembershipManager.php:
```diff
@@ -214,6 +214,11 @@ public function getGroupIds(EntityInterface $entity, $group_type_id = NULL, $gro
 
+    // This should only be called on group content types.
+    if (!$this->groupAudienceHelper->hasGroupAudienceField($entity->getEntityTypeId(), $entity->bundle())) {
+      throw new \InvalidArgumentException('Can only retrieve group IDs for group content entities.');
```
Can config entities belong to a group? If so I'd go for:

```suggestion
      throw new \InvalidArgumentException('Can only retrieve group IDs for group entities.');
```

A bit further (and most of the time we just refer to **group content**):

>     // Only delete group content that is fully orphaned, i.e. it is no longer
```suggestion
      throw new \InvalidArgumentException('Can only retrieve group IDs for group content.');
```

Remains this question: a user doesn't have an og_audience field but don't we consider it "group content"?

pfrenssen replied:
Since the group is referenced on a entity reference field, all group content needs to be a content entity (since config entities are not fieldable). Also the name "group content" implies that it is a content entity :grin: 

Users are not considered to be group content, they are considered to be group members, and are managed not through a reference field, but through the `OgMembership` entity tables.

MPParsley replied:
Note that webform is a bundlable config entity and has plans to make it fieldable (not that I would suggest doing this).
See https://www.drupal.org/project/webform/issues/3049023#comment-13073385

MPParsley commented on src/OgAccess.php:
```diff
@@ -262,36 +262,37 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+      $cache_tags = $entity_type->getListCacheTags();
+
+      // The entity might be a user or a non-user entity.
+      $groups = $entity->getEntityTypeId() == 'user' ? $this->membershipManager->getUserGroups($entity) : $this->membershipManager->getGroups($entity);
```
Or we could test for Drupal\user\UserInterface?

Also use === of you want strict typing

Also might have been useful to have this check within getGroups().

pfrenssen replied:
I agree with this proposal but these lines of code are not in fact being changed here, this PR just introduces this single line and moves the related code in it:

```
if ($is_group_content) {
  // All code in here is in fact unchanged, but shows up in Github as a huge chunk of changed code.
}
```

I will create a separate PR to implement this suggestion, so this can stay on scope.

pfrenssen committed:
Apply suggestions from code review

Co-Authored-By: Maarten Segers <MPParsley@users.noreply.github.com>
